src/components/Item/VisualizationItem/ItemContextMenu/__tests__/ItemContextMenu.offline.spec.js

```diff
@@ -0,0 +1,479 @@
+        <WindowDimensionsProvider>
+            <ItemContextMenu {...defaultProps} />
+        </WindowDimensionsProvider>
+    )
+    fireEvent.click(getByRole('button'))
+
+    await waitFor(() => {
+        expect(queryByText('View fullscreen')).toBeNull()
+    })
+})
+
+test('does not render "Show interpretations" option if settings do not allow', async () => {
+    useSystemSettings.mockImplementation(() => ({
+        settings: Object.assign({}, mockSystemSettingsDefault.settings, {
+            allowVisShowInterpretations: false,
+        }),
+    }))
+
+    const { getByRole, queryByText } = render(
+        <WindowDimensionsProvider>
+            <ItemContextMenu {...defaultProps} />
+        </WindowDimensionsProvider>
+    )
+    fireEvent.click(getByRole('button'))
+
+    await waitFor(() => {
+        expect(queryByText('Show details and interpretations')).toBeNull()
+    })
+})
+
+test('renders null if all relevant settings are false', async () => {
+    useSystemSettings.mockImplementation(() => ({
+        settings: {
+            allowVisOpenInApp: false,
+            allowVisShowInterpretations: false,
+            allowVisViewAs: false,
+            allowVisFullscreen: false,
+        },
+    }))
+
+    const { container } = render(
+        <WindowDimensionsProvider>
+            <ItemContextMenu {...defaultProps} />
+        </WindowDimensionsProvider>
+    )
+
+    expect(container).toMatchSnapshot()
+})
+
+test('renders correct options for BAR in small screen', async () => {
+    global.innerWidth = 480
+    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
+
+    const { getByRole, queryByTestId, queryByText } = render(
+        <WindowDimensionsProvider>
+            <ItemContextMenu {...defaultProps} />
+        </WindowDimensionsProvider>
+    )
+
+    fireEvent.click(getByRole('button'))
+
+    await waitFor(() => {
+        expect(queryByText('View as Map')).toBeTruthy()
+        expect(queryByText('View as Chart')).toBeNull()
+        expect(queryByText('View as Table')).toBeTruthy()
+        expect(queryByTestId('divider')).toBeTruthy()
+        expect(queryByText('Open in Data Visualizer app')).toBeNull()
+        expect(queryByText('Show details and interpretations')).toBeTruthy()
+        expect(queryByText('View fullscreen')).toBeTruthy()
+    })
+
+    global.innerWidth = 800
+})
+
+test('renders correct options for PIE in small screen', async () => {
+    global.innerWidth = 480
+    useSystemSettings.mockImplementation(() => mockSystemSettingsDefault)
+
+    const props = Object.assign({}, defaultProps, {
+        visualization: {
+            type: 'PIE',
+        },
+    })
+
+    const { getByRole, queryByTestId, queryByText } = render(
+        <WindowDimensionsProvider>
+            <ItemContextMenu {...props} />
+        </WindowDimensionsProvider>
+    )
+
+    fireEvent.click(getByRole('button'))
+
+    await waitFor(() => {
+        expect(queryByText('View as Map')).toBeNull()
```

Comment:
Since these are repeated across each step which makes the file pretty big, maybe they could be abstracted with a function like `expectViewAsMap = state => state ? expect(queryByText('View as Map')).toBeTruthy() : expect(queryByText('View as Map')).toBeNull()` to avoid repetition, or maybe even use a config for the tests, like `pieSmallScreen = { showViewAsMap: false, showViewAsChart: false, ... }`. To make it a bit easier to overview and maintain 🙏

Reply:
I have usually done things the way you describe, but I read an article not too long ago about how to organize tests, and it recommended not pulling out common code into separate functions. Trade off is a lot of repetition, but on the other hand, it is much easier to understand the each test. That said, I'm not sure I'm in 100% agreement with the author. I'll try to find the article.